nemo_curator/modules/fuzzy_dedup.py

```diff
@@ -1652,52 +1666,52 @@ def _write_encoded_jaccard_pair(self, dedup_parsed_id_path):
                 id_columns = ["dataset_id", "doc_id"]
 
             num_workers = get_num_workers(get_current_client())
-            self._batched_merge_and_write(
+            self._merge_and_write(
                 ddf=ddf,
                 ddf_id=ddf_id,
                 output_path=output_path,
                 id_columns=id_columns,
-                batch_size=num_workers,
+                # batch_size=num_workers,
             )
         self._logger.info(
             f"Time taken for Encoding Jaccard Pairs = {time.time() - t0}s and output written at {output_path}"
         )
         return output_path
 
-    def _batched_merge_and_write(
-        self, ddf, ddf_id, output_path, id_columns, batch_size=32
-    ):
-        total_batches = (ddf.npartitions + batch_size - 1) // batch_size
-        for batch_id, offset in enumerate(range(0, ddf.npartitions, batch_size)):
-            st = time.time()
-            subset_ddf = ddf.partitions[offset : offset + batch_size]
-            for tag in ["x", "y"]:
-                pair_ids = []
-                for id_col in id_columns:
-                    pair_ids.append(f"{id_col}_{tag}")
-                subset_ddf = subset_ddf.merge(
-                    ddf_id,
-                    left_on=pair_ids,
-                    right_on=id_columns,
-                    how="inner",
-                    broadcast=True,
-                )
-                subset_ddf = subset_ddf.drop(
-                    columns=pair_ids,
-                )
-                subset_ddf = subset_ddf.rename(
-                    columns={"uid": f"{self.id_column}_{tag}"}
-                )
+    def _merge_and_write(self, ddf, ddf_id, output_path, id_columns):
+        import time
 
-            subset_ddf = subset_ddf[[self.left_id, self.right_id, "jaccard"]]
-            output_batch_path = os.path.join(output_path, f"{batch_id}.parquet")
-            subset_ddf.to_parquet(output_batch_path, write_index=False)
+        st = time.time()
 
-            et = time.time()
-            print(
-                f"batch_id = {batch_id}/{total_batches}, time = {et - st}", flush=True
+        # Ensure 'id_columns' is a list
+        if isinstance(id_columns, str):
+            id_columns = [id_columns]
+
+        # Optimize 'ddf_id' by setting 'id_columns' as the index
+        ddf_id = ddf_id.set_index(id_columns)
+
+        # Perform the merge operations for 'x' and 'y' tags
+        for tag in ["x", "y"]:
+            pair_ids = [f"{id_col}_{tag}" for id_col in id_columns]
+
+            # Merge 'ddf' with 'ddf_id' to map ids to uids
+            ddf = ddf.merge(
+                ddf_id,
+                left_on=pair_ids,
+                right_index=True,
+                how="inner",
+                # shuffle="p2p",
+                broadcast=True,
```

Comment:
Is there a limit / recommendation on broadcast merge, given that now `ddf_id` which I believe is the smaller of the two will be unbounded, whereas in the original implementation it is batched? I worry if it'll OOM at certain thresholds

Reply:
Discussed this with Praateek on a call.  Couple of clarifications: 

1. It was previously also a broadcast merge so nothing changes there. 
2. We have 2*num_batches -1 shuffles with this index once and forget approach 
3. The only extra set is set_index on `ddf_id` column